src/Storages/MergeTree/MergeTreeIndexMinMax.cpp

```diff
@@ -72,29 +56,53 @@ void MergeTreeIndexGranuleMinMax::deserializeBinary(ReadBuffer & istr)
         const DataTypePtr & type = index_sample_block.getByPosition(i).type;
         auto serialization = type->getDefaultSerialization();
 
-        if (!type->isNullable())
-        {
-            serialization->deserializeBinary(min_val, istr);
-            serialization->deserializeBinary(max_val, istr);
-        }
-        else
+        switch (version)
         {
-            /// NOTE: that this serialization differs from
-            /// IMergeTreeDataPart::MinMaxIndex::load() due to preserve
-            /// backward compatibility.
-            bool is_null;
-            readBinary(is_null, istr);
-            if (!is_null)
-            {
+            case 1:
+                if (!type->isNullable())
+                {
+                    serialization->deserializeBinary(min_val, istr);
+                    serialization->deserializeBinary(max_val, istr);
+                }
+                else
+                {
+                    /// NOTE: that this serialization differs from
+                    /// IMergeTreeDataPart::MinMaxIndex::load() to preserve
+                    /// backward compatibility.
+                    ///
+                    /// But this is deprecated format, so this is OK.
+
+                    bool is_null;
+                    readBinary(is_null, istr);
+                    if (!is_null)
+                    {
+                        serialization->deserializeBinary(min_val, istr);
+                        serialization->deserializeBinary(max_val, istr);
+                    }
+                    else
+                    {
+                        min_val = Null();
+                        max_val = Null();
+                    }
+                }
+                break;
+
+            /// New format with proper Nullable support for values that includes Null values
+            case 2:
                 serialization->deserializeBinary(min_val, istr);
                 serialization->deserializeBinary(max_val, istr);
-            }
-            else
-            {
-                min_val = Null();
-                max_val = Null();
-            }
+
+                // NULL_LAST
+                if (min_val.isNull())
+                    min_val = PositiveInfinity();
```

Comment:
Ohhh so this questionable PR that added infinities into the variant type was merged...
By the way, can you serialize it as PositiveInfinity right away? Having to convert it afterwards looks weird.

Reply:
I guess we can have a small function `deserializeBinaryNullAsPositiveInfinity()` and put the logic there. These infinity varants share the same serialization bits as NULL.

Reply:
>I guess we can have a small function deserializeBinaryNullAsPositiveInfinity() and put the logic the

@amosbird You mean `MergeTreeIndexgranuleMinMax::deserializeBinaryNullAsPositiveInfinity()` right?

Reply:
Oh, I just realized that we don't use `writeBinary(Field)` here, but instead for some reason we use the data type serialization to serialize a Field, which obviously can't serialize the special Field values. Never mind, it's all equally awkward, let's just keep it as is.

Reply:
So instead we have to serialize what we can, and then guess what it was. At least it's always `NULLS LAST`, for now.